File: haskell/2021/day_10.hs
import qualified Control.Monad as Monad
import qualified Data.Char as Char
import qualified Data.Either as Either
import qualified Data.List as List
import qualified Data.Maybe as Maybe

-- | Attempt to autocomplete a string of brackets.
-- Return a corrupt score (Left x) or completion score (Right x).
autoComplete :: String -> Either Int Int
autoComplete chars = do
    -- Create a stack of unclosed brackets
    stack <- Monad.foldM (\stack c ->
        if c `elem` "([{<"
        then Right $ c:stack
        else case stack of
            -- Check matching brackets by comparing ASCII value difference
            o:rest | o < c && Char.ord c - Char.ord o < 3 -> Right rest
            _ -> Left $ case c of
                ')' -> 3
                ']' -> 57
                '}' -> 1197
                '>' -> 25137
                _ -> error "Invalid char"
        ) [] chars
    Right $ foldl (\acc c -> 5 * acc + 1 + Maybe.fromJust (c `List.elemIndex` "([{<")) 0 stack

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_10.txt"
    let (corruptScores, autoScores) = Either.partitionEithers $ autoComplete <$> lines text
    putStrLn $ "Part 1: " ++ show (sum corruptScores)
    let sorted = List.sort autoScores
    putStrLn $ "Part 2: " ++ show (sorted !! (length sorted `div` 2))

File: rust/2021/src/bin/day_10.rs
use std::fs;

fn autocomplete(line: &str) -> Result<u64, u64> {
    let mut stack = vec![];
    // Read brackets into stack
    for byte in line.bytes() {
        match byte {
            b'(' | b'[' | b'{' | b'<' => stack.push(byte),
            close => match stack.last() {
                // Check matching brackets by comparing ASCII value difference
                Some(&open) if open < close && close - open < 3 => {
                    stack.pop();
                }
                _ => {
                    return Err(match close {
                        b')' => 3,
                        b']' => 57,
                        b'}' => 1197,
                        b'>' => 25137,
                        _ => panic!("Invalid byte"),
                    })
                }
            },
        }
    }
    // Calculate auto-completion score
    Ok(stack.iter().rfold(0, |acc, &open| {
        5 * acc
            + match open {
                b'(' => 1,
                b'[' => 2,
                b'{' => 3,
                b'<' => 4,
                _ => panic!("Invalid byte"),
            }
    }))
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_10.txt").unwrap();
    let mut corrupt_total = 0;
    let mut auto_scores = vec![];
    for result in input.lines().map(autocomplete) {
        match result {
            Ok(score) => auto_scores.push(score),
            Err(score) => corrupt_total += score,
        }
    }
    auto_scores.sort();
    println!("Part 1: {}", corrupt_total);
    println!("Part 2: {}", auto_scores[auto_scores.len() / 2])
}
